graph-gateway/src/network/entities.rs

```diff
@@ -0,0 +1,192 @@
+//! Entities that are used to represent the network topology.
+
+use std::{
+    collections::{HashMap, HashSet},
+    fmt::Display,
+    ops::Deref,
+    sync::Arc,
+};
+
+pub use alloy_primitives::{Address, BlockNumber};
+use cost_model::CostModel;
+use custom_debug::CustomDebug;
+use eventuals::Ptr;
+use semver::Version;
+pub use thegraph_core::types::{DeploymentId, SubgraphId};
+use url::Url;
+
+/// The [`IndexingId`] struct represents the unique identifier of an indexing.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct IndexingId {
+    /// The indexer's ID indexing the subgraph's deployment.
+    pub indexer: Address,
+    /// The subgraph's deployment ID indexed by the indexer.
+    pub deployment: DeploymentId,
+}
+
+#[derive(Clone)]
+pub struct Indexing {
+    /// The indexing unique identifier.
+    pub id: IndexingId,
+
+    /// The versions behind the highest version of the subgraph being indexed.
+    pub versions_behind: u8,
+
+    /// The indexer
+    pub indexer: Arc<Indexer>,
+
+    /// The indexer's indexing status
+    pub status: Option<IndexingStatus>,
+    /// The indexer's indexing cost model
+    pub cost_model: Option<Ptr<CostModel>>,
+}
+
+/// The [`IndexingStatus`] struct represents the indexer's indexing status.
+#[derive(Debug, Clone)]
+pub struct IndexingStatus {
+    /// The latest block the indexer has indexed for the deployment.
+    pub latest_block: BlockNumber,
+    /// The minimum block the indexer has indexed for the deployment.
+    pub min_block: Option<BlockNumber>,
+}
+
+/// The [`Indexer`] struct represents an indexer in the network topology.
+///
+/// The indexer is a network  node that indexes a subgraph using one of its versions, i.e., one of
+/// the subgraph's deployments. The [`Indexing`] struct represents the indexer's indexing of a
+/// subgraph's deployment.
+#[derive(CustomDebug, Clone)]
+pub struct Indexer {
+    /// The indexer's ID.
+    pub id: Address,
+
+    /// The indexer's URL.
+    ///
+    /// It is guaranteed that the URL scheme is either HTTP or HTTPS and the URL has a host.
+    #[debug(with = Display::fmt)]
+    pub url: Url,
+
+    /// The total amount of tokens staked by the indexer.
+    pub staked_tokens: u128,
+    /// The largest allocation address.
+    pub largest_allocation: Address,
```

Comment:
This field is concerning. The allocation should be associated with an indexing, not with the indexer.

Reply:
I will review this 👍🏼

Reply:
I did my best trying to untangle the original spaghetti code, but seems I missed one turn 😅 🍝 

I pushed the fix in [`348f0fd` (#706)](https://github.com/edgeandnode/gateway/pull/706/commits/348f0fdd3f219d97c3f7cc88ad80c38dc51adb9d)